src/OpenLoco/src/Drawing/SoftwareDrawingContext.cpp

```diff
@@ -1537,17 +1537,129 @@ namespace OpenLoco::Drawing
             fillRectInset(rt, x, y, x + dx - 1, y + dy - 1, colour, flags);
         }
 
+        /**
+         * Draws a horizontal line of specified colour to a buffer.
+         *  0x0045308A
+         */
+        static void drawHorizontalLine(Gfx::RenderTarget& rt, PaletteIndex_t colour, const Ui::Point& startCoord, int32_t length)
+        {
+            Ui::Point offset(startCoord.x - rt.x, startCoord.y - rt.y);
+
+            // Check to make sure point is in the y range
+            if (offset.y < 0)
+                return;
+            if (offset.y >= rt.height)
+                return;
+            // Check to make sure we are drawing at least a pixel
+            if (length == 0)
+                length++;
+
+            // If x coord outside range leave
+            if (offset.x < 0)
+            {
+                // Unless the number of pixels is enough to be in range
+                length += offset.x;
+                if (length <= 0)
+                    return;
+                // Resets starting point to 0 as we don't draw outside the range
+                offset.x = 0;
+            }
+
+            // Ensure that the end point of the line is within range
+            if (offset.x + length - rt.width > 0)
+            {
+                // If the end point has any pixels outside range
+                // cut them off. If there are now no pixels return.
+                length -= offset.x + length - rt.width;
+                if (length <= 0)
+                    return;
+            }
+
+            // Get the buffer we are drawing to and move to the first coordinate.
+            uint8_t* buffer = rt.bits
+                + offset.y * static_cast<int32_t>(rt.pitch + rt.width) + offset.x;
+
+            // Draw the line to the specified colour
+            std::fill_n(buffer, length, colour);
+        }
+
         // 0x00452DA4
-        static void drawLine(Gfx::RenderTarget& rt, int16_t left, int16_t top, int16_t right, int16_t bottom, uint32_t colour)
+        void drawLine(Gfx::RenderTarget& rt, int16_t left, int16_t top, int16_t right, int16_t bottom, PaletteIndex_t colour)
         {
-            registers regs;
-            regs.ax = left;
-            regs.bx = top;
-            regs.cx = right;
-            regs.dx = bottom;
-            regs.ebp = colour;
-            regs.edi = X86Pointer(&rt);
-            call(0x00452DA4, regs);
+            // Check to make sure the line is within the drawing area
+            if ((left < rt.x) && (right < rt.x))
+            {
+                return;
+            }
+
+            if ((top < rt.y) && (bottom < rt.y))
+            {
+                return;
+            }
+
+            if ((left > (rt.x + rt.width)) && (right > (rt.x + rt.width)))
+            {
+                return;
+            }
+
+            if ((top > (rt.y + rt.height)) && (bottom > (rt.y + rt.height)))
+            {
+                return;
+            }
```

Comment:
I think we have a helper function for this in `Rect::intersects`. We should probably change the parameter to a `Rect`, adding a compat function for the old signature. Thoughts?

Reply:
ah good point.

Reply:
I decided that giving a Rect as input to a line draw function would be weird. I've gone for two Ui::Points a and b. I then construct a Rect to do the intersect.